### PR TITLE
Project Completed.

### PR DESCRIPTION
**Pull request by AshtonLeeSeloka**

### CODE REVIEW

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
+﻿using CodingTracker.AshtonLeeSeloka.Models;
+using CodingTracker.AshtonLeeSeloka.Views;
+using Services;
+using Spectre.Console;
+using static CodingTracker.AshtonLeeSeloka.Models.MenuItems;
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:7]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+				case MenuInsert.Manually_Record_Session:
+					ManualInsert();
+					break;
+				case MenuInsert.Start_Timer:
+					TimerInsert();
+					break;
+				case MenuInsert.Back:
+					break;
+			}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+			while (startDate > endDate)
+			{
+				Console.WriteLine("End date value cannot be earlier than start date value, press any key to re-enter session end value");
+				Console.ReadKey();
+				endDate = _validation.GetValidatedDate("end");
+			}
+
+			float time = _calculationsService.GetDuration(startDate, endDate);
+			_dataService.Insert(startDate.ToString(), endDate.ToString(), (float)System.Math.Round(time, 2));
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs]: Unnecessary empty lines.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+					DateDescendingSessions(codingSessions);
+					break;
+				case MenuViewData.View_in_ascending_order:
+
+					DateAscendingSession(codingSessions);
+					break;
+				case MenuViewData.Back:
+					break;
+			}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:78]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		}
+
+		public void DateAscendingSession(List<CodingSession> codingSessions)
+		{
+			codingSessions.Sort((x, y) => x.StartTime.CompareTo(y.StartTime));
+			_view.DisplaySessionView(codingSessions);
+			Console.WriteLine("Press any key to return");
+			Console.ReadKey();
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:100]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+		}
+
+		public void DateDescendingSessions(List<CodingSession> codingSessions)
+		{
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:115]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+			codingSessions.Sort((x, y) => DateTime.Parse(y.StartTime).CompareTo(DateTime.Parse(x.StartTime)));
+			//codingSessions.OrderByDescending(x => DateTime.Parse(x.StartTime));
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs]: Unused code

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+			_view.DisplaySessionView(codingSessions);
+			Console.WriteLine("Press any key to return");
+			Console.ReadKey();
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:122]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+			switch (selection)
+			{
+				case MenuItems.MenuReport.View_All_Data:
+					ReportAllSessions();
+					break;
+				case MenuItems.MenuReport.Specify_Report_Range:
+					ReportRangeSessions();
+					break;
+			}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:153]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+			for (int i = 0; i < session.Count; i++)
+			{
+				DateTime iStart = DateTime.Parse(session[i].StartTime);
+				DateTime iEnd = DateTime.Parse(session[i].StartTime);
+
+				if (iStart < startDate || iEnd > endDate)
+				{
+					session.RemoveAt(i);
+				}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs]: Unnecessary empty lines

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+			session.Sort((x, y) => x.StartTime.CompareTo(y.StartTime));
+
+			_view.DisplaySessionView(session);
+			float? sum = _calculationsService.CalculateSum(session);
+			float? avg = _calculationsService.CalculateAverage(session);
+			int entries = session.Count();
+			string? startDateRange = session[0].StartTime;
+			string? endDateRange = session[(entries - 1)].EndTime;
+			_view.DisplayAllReportDataView(startDateRange, endDateRange, entries, avg, sum);
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs]: Unnecessary empty lines.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+		}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Controllers/CodingController.cs:177]: Unnecessary empty lines

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Enums/MenuItems.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+		View_All_Data,
+		Back
+	}
+
+	public enum MenuViewData
+	{
+		View_in_ascending_order,
+		View_in_descending_order,
+		Back
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Enums/MenuItems.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Enums/MenuItems.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Enums/MenuItems.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+﻿using CodingTracker.AshtonLeeSeloka.UserInterface;
+using Services;
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs:3]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+
+//Creating DB IF not Present
+DataService _DB =  new DataService();
+_DB.CreateDB();
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿using CodingTracker.AshtonLeeSeloka.UserInterface;
+using Services;
+
+
+//Creating DB IF not Present
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Program.cs]: Unnecessary comment

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/UserInterface.cs [hunk 18]
```diff
@@ -0,0 +1,47 @@
+﻿using CodingTracker.AshtonLeeSeloka.Controllers;
+using Spectre.Console;
+using static CodingTracker.AshtonLeeSeloka.Models.MenuItems;
+namespace CodingTracker.AshtonLeeSeloka.UserInterface;
+
+internal class UserInterface
+{
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/UserInterface.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/UserInterface.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+	public void MainMenu()
+	{
+		CodingController controller = new CodingController();
+
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/UserInterface.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+				table.AddRow(
+					$"[cyan]{codingSession.Id.ToString()}[/]",
+					$"[yellow] {codingSession.StartTime}[/]",
+					$"[yellow] {codingSession.EndTime}[/]",
+					$"[yellow] {codingSession.Duration}[/]"
+				);
+			}
+
+			AnsiConsole.Write(table);
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs:44]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+		public MenuItems.MenuReport ReportMenuView()
+		{
+			var selectedOption = AnsiConsole.Prompt(
+				new SelectionPrompt<MenuReport>()
+				.Title("[green]Select Reporting option[/]")
+				.AddChoices(Enum.GetValues<MenuReport>())
+				);
+			return selectedOption;
+		}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs:105]: Unnecessary empty lines

File: CodingTracker.AshtonLeeSeloka/Services/CalculationsService.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+		{
+			float? sum = 0;
+			foreach (CodingSession sess in sessons)
+			{
+				sum = sum + sess.Duration;
+			}
+
+			return sum;
+		}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/CalculationsService.cs:27]: Four unnecessary empty lines!

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 23]
```diff
@@ -0,0 +1,133 @@
+﻿using CodingTracker.AshtonLeeSeloka.Models;
+using Dapper;
+using Microsoft.Data.Sqlite;
+using System.Configuration;
+
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+namespace Services
+{
+	public class DataService
+	{
+		private string? _DBConnectionString = ConfigurationManager.AppSettings.Get("DBConnectionString");
+		private List<CodingSession> _Sessions = new List<CodingSession>();
+		private CalculationsService _CalculationsService = new CalculationsService();
+		private ValidationService _Validation = new ValidationService();
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+		public void CreateDB()
+		{
+			try
+			{
+				using (var connection = new SqliteConnection(_DBConnectionString))
+				{
+					connection.Open();
+					var tableCmd = connection.CreateCommand();
+					tableCmd.CommandText = @"CREATE TABLE IF NOT EXISTS Coding_Sessions(
+																Id INTEGER PRIMARY KEY AUTOINCREMENT,
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs:26]: Three unnecessary empty lines! Why? 😄

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+																Duration(Hours) FLOAT)";
+					tableCmd.ExecuteNonQuery();
+				}
+			}
+			catch (Exception ex)
+			{
+				Console.Clear();
+				Console.WriteLine(ex.Message);
+				Console.WriteLine("Type 0 to exit");
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+			}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+		/// <summary>
+		/// Method to Delete selected CodingSession object
+		/// </summary>
+		/// <param name="session">Object to be deleted</param>
+		public void Delete(CodingSession session)
+		{
+			var sqlCommand = "DELETE FROM coding_Sessions WHERE Id = @ID";
+			var connection = new SqliteConnection(_DBConnectionString);
+			connection.Execute(sqlCommand, new { ID = session.Id });
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs:40]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+
+			Console.WriteLine("\nDeletion of Coding session succesful, Type any Key to exit");
+			Console.ReadLine();
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs:68]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+			Console.ReadLine();
+		}
+
+		/// <summary>
+		/// Gets all sessions in the form of List<CodingSession>
+		/// </summary>
+		/// <returns>List<CodingSession></returns>
+		public List<CodingSession> GetAllSessions()
+		{
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+			_Sessions.Clear();
+
+			var sqlCommand = "SELECT * FROM Coding_Sessions";
+			var connection = new SqliteConnection(_DBConnectionString);
+			var codingSessions = connection.Query<CodingSession>(sqlCommand);
+
+			foreach (var codingSession in codingSessions)
+			{
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+				{
+					Id = codingSession.Id,
+					StartTime = codingSession.StartTime,
+					EndTime = codingSession.EndTime,
+					Duration = codingSession.Duration,
+				});
+			}
+
+			return _Sessions;
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/DataService.cs]: Unnecessary empty line.

File: CodingTracker.AshtonLeeSeloka/Services/ValidationService.cs [hunk 33]
```diff
@@ -0,0 +1,56 @@
...
+				{
+					Console.Clear();
+					Console.WriteLine("Please enter date in correct format (yyyy-MM-dd HH:mm:ss)");
+					Console.WriteLine("Press Any key to retry)");
+					Console.ReadLine();
+				}
+			}
+			return validatedDate;
+		}
+
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/Services/ValidationService.cs]: Four unnecessary empty lines!

File: CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs [hunk 34]
```diff
@@ -0,0 +1,133 @@
+﻿using CodingTracker.AshtonLeeSeloka.Models;
+using Spectre.Console;
+using System;
```
Comment (TheCSharpAcademy) [inline CodingTracker.AshtonLeeSeloka/CodingTracker.AshtonLeeSeloka/Views/View.cs]: Unused usings

**Review by TheCSharpAcademy — CHANGES_REQUESTED** [review]
@AshtonLeeSeloka Well done! It's looking great! Before we approve could you please tidy up according to comments? [You can read our code conventions here](https://www.thecsharpacademy.com/article/58/code-conventions).

File: CodingTracker.AshtonLeeSeloka/Services/DataService.cs (hunk 25 above)
Comment (AshtonLeeSeloka) [reply CodingTracker.AshtonLeeSeloka/Services/DataService.cs:26]: Lol Apologies, will be more mindful in future,
I owe you a coffee one day for the patience lol.

I've resubmitted for your review.

**Comment by TheCSharpAcademy** [comment]
@AshtonLeeSeloka  Project approved! 😄✅ And well done for completing some of the challenges! Especially the reports. I suggest you always push the boundaries of your SQL knowledge because that's pretty much what the internet is all about.

🗺️And well done! You got your 🟩🟩 Olive Green Belt!

🧑🏻‍🏫In terms of feedback, make sure you clean up a bit before submitting as we tend to get more strict with code tidiness. 🕵🏻‍♀️ Also, this is a personal choice but we think a monolith works just fine for most projects in the academy, so it's unnecessary to create several projects, unless you're practicing architectural patterns such as Clean or N-Tier.

✋🏻The next project will be a step up in terms of complexity, but we recommend you do the challenge to learn about the "Pivot" keyword, a very important skill in your SQL toolbox.🫡

**Pull request closed without merging**